src/server/mod.rs

```diff
@@ -110,31 +117,51 @@ impl ClientSource {
     }
 }
 
+/// A complete MQTT Server
+///
+/// This server should be seen as a toolkit to integrate with your application.
+///
+/// To use it, you first need to get a new instance of it, check out any of the `serve_*` methods
+/// that create a new instance.
+/// Then, you need to start listening for new connections in a long lived future.
+///
+/// Check out the server example for a working version.
+///
 pub struct MqttServer {
-    clients: DashMap<ClientId, ClientState>,
-    client_source: ClientSource,
+    clients: Arc<DashMap<ClientId, ClientState>>,
+    client_source: Mutex<ClientSource>,
     subscription_manager: SubscriptionManager,
 }
 
 impl MqttServer {
+    /// Create a new MQTT server listening on the given `SocketAddr`
     pub async fn serve_v3_unsecured_tcp<Addr: ToSocketAddrs>(
         addr: Addr,
     ) -> Result<Self, MqttError> {
         let bind = TcpListener::bind(addr).await?;
 
         Ok(MqttServer {
-            clients: DashMap::new(),
-            client_source: ClientSource::UnsecuredTcp(bind),
+            clients: Arc::new(DashMap::new()),
+            client_source: Mutex::new(ClientSource::UnsecuredTcp(bind)),
             subscription_manager: SubscriptionManager::new(),
         })
     }
 
-    pub async fn accept_new_clients(&mut self) -> Result<(), MqttError> {
+    /// Start accepting new clients connecting to the server
+    pub async fn accept_new_clients(self: Arc<Self>) -> Result<(), MqttError> {
+        let mut client_source = self
+            .client_source
+            .try_lock()
+            .map_err(|_| MqttError::AlreadyListening)?;
+
         loop {
-            let client = self.client_source.accept().await?;
-            if let Err(client_error) = self.accept_client(client).await {
-                tracing::error!("Client error: {}", client_error)
-            }
+            let client = client_source.accept().await?;
+            let server = self.clone();
+            tokio::spawn(async move {
+                if let Err(client_error) = server.accept_client(client).await {
+                    tracing::error!("Client error: {}", client_error)
+                }
+            });
         }
```

Comment:
Would it make sense to have the function return a `Stream<Item = Result<(), MqttError>` and transform this bit of code into a stream?
This way, the user could decide how to act upon errors.

Reply:
I don't think so, this is meant to be used as a pre-packaged server and as such is really self-contained. 

We could at some point include some form of 'callbacks' or so to have some places one could hook into, but other than that I think its better to see this as its own little island.